File: src/data/courses/dsa/merge-sort/sorting-comparisons/space-complexity-analysis/code/python/space_complexity_analysis.py
# Merge Sort: O(n) space
def merge(arr, l, m, r):
    L = arr[l:m+1]  # O(n) space
    R = arr[m+1:r+1]
    # ... merge logic

# Quick Sort: O(log n) space (recursion)
def quick_sort(arr, l, r):
    if l < r:
        p = partition(arr, l, r)
        quick_sort(arr, l, p - 1)
        quick_sort(arr, p + 1, r)

# Heap Sort: O(1) space
def heap_sort(arr):
    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)

File: src/data/courses/dsa/merge-sort/sorting-comparisons/space-complexity-analysis/code/java/SpaceComplexityAnalysis.java
// Merge Sort: O(n) space
void merge(int[] arr, int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    int[] L = new int[n1], R = new int[n2]; // O(n) space
    // ... merge logic
}

// Quick Sort: O(log n) space (recursion stack)
void quickSort(int[] arr, int l, int r) {
    if (l < r) {
        int p = partition(arr, l, r); // O(1) space
        quickSort(arr, l, p - 1);     // Stack depth O(log n)
        quickSort(arr, p + 1, r);
    }
}

// Heap Sort: O(1) space
void heapSort(int[] arr) {
    int n = arr.length;
    for (int i = n / 2 - 1; i >= 0; i--) heapify(arr, n, i);
    for (int i = n - 1; i > 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;
        heapify(arr, i, 0);
    }
}